tests/test_delineateit.py

```diff
@@ -345,3 +349,134 @@ def test_check_geometries(self):
+        # buffered with a border of nodata
+        flow_dir_array = numpy.array([
+            [7, 7, 7, 5],
+            [6, 6, 6, 4],
+            [0, 1, a, 0],
+            [a, 2, 3, 4]
+        ], dtype=numpy.intc)
+
+        edges = numpy.array([1, 1, 0, 0], dtype=numpy.intc)  # top, left, bottom, right
+
+        expected_pour_points = {(2, 1)}
+
+        output = delineateit_core.calculate_pour_point_array(flow_dir_array, edges, a)
+        self.assertTrue(numpy.array_equal(
+            output,
+            expected_pour_points))
+
+    def test_expand_and_find_edges(self):
+        from natcap.invest.delineateit import delineateit
+
+        raster_size = (10, 4)
+        blocks = [
+            {'xoff': 0, 'yoff': 0, 'win_xsize': 4, 'win_ysize': 4},
+            {'xoff': 4, 'yoff': 0, 'win_xsize': 4, 'win_ysize': 4},
+            {'xoff': 8, 'yoff': 0, 'win_xsize': 2, 'win_ysize': 4}
+        ] 
+
+        expected_blocks = [
+            {'xoff': 0, 'yoff': 0, 'win_xsize': 5, 'win_ysize': 4},
+            {'xoff': 3, 'yoff': 0, 'win_xsize': 6, 'win_ysize': 4},
+            {'xoff': 7, 'yoff': 0, 'win_xsize': 3, 'win_ysize': 4}
+        ]
+        expected_edges = [
+            [1, 1, 1, 0],  # top, left, bottom, right
+            [1, 0, 1, 0],
+            [1, 0, 1, 1]
+        ]
+
+        for in_block, expected_block, expected_edges in zip(blocks, 
+                                                            expected_blocks, 
+                                                            expected_edges):
+            block, edges = delineateit._expand_and_find_edges(in_block, raster_size)
+            self.assertTrue(numpy.array_equal(block, expected_block))
+            self.assertTrue(numpy.array_equal(edges, expected_edges))
+
+    def test_find_pour_points_by_block(self):
+        from natcap.invest.delineateit import delineateit
+
+        a = 100  # nodata value
+        flow_dir_array = numpy.array([
+            [0, 0, 0, 0, 7, 7, 7, 1, 6, 6],
+            [2, 3, 4, 5, 6, 7, 0, 1, 1, 2],
+            [2, 2, 2, 2, 0, a, a, 3, 3, a],
+            [2, 1, 1, 1, 2, 6, 4, 1, a, a],
+            [1, 1, 0, 0, 0, 0, a, a, a, a]
+        ], dtype=numpy.int8)
+
+        raster_info = {
+            'raster_size': (10, 4), 
+            'pixel_size': (1, 1), 
+            'nodata': (a,),
+            'geotransform': [0, 0, 0, 0, 0, 0]}
+
+        srs = osr.SpatialReference()
+        srs.ImportFromEPSG(3157)
+        projection_wkt = srs.ExportToWkt()
+
+        raster_path = os.path.join(self.workspace_dir, 'small_raster.tif')
+        pygeoprocessing.numpy_array_to_raster(
+            flow_dir_array,
+            a,
+            (1, 1),
+            (0, 0),
+            projection_wkt,
+            raster_path
+        )
+
+        expected_pour_points = {(7.5, 0.5), (5.5, 1.5), (4.5, 2.5), (5.5, 4.5)}
+
+        # Mock iterblocks so that we can test with an array smaller than 128x128

```

Comment:
Is this just to make sure that we can operate on smaller block sizes than the default?  If so, can't we just set the GDAL raster creation options directly using a smaller power-of-2 number and then use `iterblocks` directly?

Reply:
This was just to have a small, easy-to-understand test case, so that I could include the numpy array written out in the test, and you can easily tell which points are supposed to be pour points.
I didn't realize you could change the block size when creating the raster; I'll try that and see if I can do away with the `mock.patch`.

Reply:
Update: changing the block size works, but `iterblocks` defaults to expanding the block size up to the maximum, which is desirable in real life but not in the test (it would all be one block, and part of the point of this test case was to ensure that pour points on the edges between blocks are handled correctly). So I would still need to patch `iterblocks` to override the `largest_block` value. Maybe this is overkill... what do you think?

Reply:
I actually don't think this is necessarily overkill.  Although `pygeoprocessing.iterblocks` can use strict block boundaries (`iterblocks(..., largest_block=0)`), I believe that GDAL itself has a lower limit on how small the blocks in a GeoTiff are allowed to be at creation time.  I don't think you'd be able to do blocks of size 4, for example, as you have here in the loop.

So maybe it'd be worth playing with the `BLOCKXSIZE` and `BLOCKYSIZE` gtiff creation options and seeing if there's something appropriate for this test?  If not, then `mock.patch` would be a reasonable workaround.